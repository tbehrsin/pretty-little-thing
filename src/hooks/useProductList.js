import React, { useState, useEffect, createContext, useContext } from 'react';
import PropTypes from 'prop-types';
import * as CustomPropTypes from 'utils/CustomPropTypes';
import { PRETTYLITTLETHING_API_URL } from '@env';

const ProductListContext = createContext();

const ProductListProvider = ({ url, children }) => {
  const [productList, setProductList] = useState();
  const [loading, setLoading] = useState(true);
  const [error, setError] = useState();

  useEffect(() => {
    let cancel = false;

    (async () => {
      const response = await fetch(`${url}/products`);
      /* istanbul ignore next */
      if (cancel) {
        return;
      }

      const productList = await response.json();
      /* istanbul ignore next */
      if (cancel) {
        return;
      }

      setProductList(productList);
      setLoading(false);
      setError();
    })().catch(err => {
      if (!cancel) {
        setProductList();
        setLoading(false);
        setError(err);
      }
    });

    return () => (cancel = true);
  }, []);

  const context = { loading, error, productList };

  return (
    <ProductListContext.Provider value={context}>
      {children}
    </ProductListContext.Provider>
  );
};

ProductListProvider.propTypes = {
  url: PropTypes.string.isRequired,
  children: PropTypes.node,
};

const MockProductListProvider = ({
  productList,
  loading = false,
  children,
}) => {
  const context = { loading, productList: loading ? undefined : productList };

  return (
    <ProductListContext.Provider value={context}>
      {children}
    </ProductListContext.Provider>
  );
};

MockProductListProvider.propTypes = {
  productList: PropTypes.arrayOf(CustomPropTypes.product),
  loading: PropTypes.bool,
  children: PropTypes.node,
};

const withProductListProvider = Component => props =>
  (
    <ProductListProvider url={PRETTYLITTLETHING_API_URL}>
      <Component {...props} />
    </ProductListProvider>
  );

const useProductList = () => useContext(ProductListContext);

export {
  ProductListProvider,
  MockProductListProvider,
  withProductListProvider,
  useProductList,
};
