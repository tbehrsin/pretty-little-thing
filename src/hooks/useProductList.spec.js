import React from 'react';
import { renderHook, act } from '@testing-library/react-hooks';
import { ProductListProvider, useProductList } from './useProductList';
import products from 'fixtures/products.json';

global.fetch = jest.fn(() =>
  Promise.resolve({
    json: () => Promise.resolve(products),
  })
);

const immediate = () => new Promise(resolve => setImmediate(resolve));

describe('useProductList', () => {
  beforeEach(() => {
    fetch.mockClear();
  });

  test('fetches product list', async () => {
    const wrapper = ({ children }) => (
      <ProductListProvider url="">{children}</ProductListProvider>
    );
    const { result } = renderHook(() => useProductList(), { wrapper });

    expect(result.current).toMatchSnapshot();

    await act(immediate);

    expect(result.current).toMatchSnapshot();
  });

  test('fetch error', async () => {
    global.fetch.mockImplementation(() => Promise.reject('API error'));

    const wrapper = ({ children }) => (
      <ProductListProvider url="">{children}</ProductListProvider>
    );
    const { result } = renderHook(() => useProductList(), { wrapper });

    expect(result.current).toMatchSnapshot();

    await act(immediate);

    expect(result.current).toMatchSnapshot();
  });
});
