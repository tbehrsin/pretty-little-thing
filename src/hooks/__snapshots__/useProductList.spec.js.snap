// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`useProductList fetch error 1`] = `
Object {
  "error": undefined,
  "loading": true,
  "productList": undefined,
}
`;

exports[`useProductList fetch error 2`] = `
Object {
  "error": "API error",
  "loading": false,
  "productList": undefined,
}
`;

exports[`useProductList fetches product list 1`] = `
Object {
  "error": undefined,
  "loading": true,
  "productList": undefined,
}
`;

exports[`useProductList fetches product list 2`] = `
Object {
  "error": undefined,
  "loading": false,
  "productList": Array [
    Object {
      "colour": "Black",
      "id": 1,
      "img": "http://cdn-img.prettylittlething.com/9/0/a/a/90aa90903a135ee59594f47c7685aa7ef3046e44_cly8063_1.jpg?imwidth=1024",
      "name": "Black Sheet Strappy Textured Glitter Bodycon Dress",
      "price": 10,
    },
    Object {
      "colour": "Stone",
      "id": 2,
      "img": "https://cdn-img.prettylittlething.com/3/6/5/a/365a5d1dce6a2b77b564379b302c9d83afccf33b_cmd2051_1.jpg?imwidth=1024",
      "name": "Stone Ribbed Strappy Cut Out Detail Bodycon Dress",
      "price": 4,
    },
    Object {
      "colour": "Black",
      "id": 3,
      "img": "https://cdn-img.prettylittlething.com/d/c/3/3/dc337260f9ecefdb99a8c8e98cd73ccb1b79cea5_cmb6804_4.jpg?imwidth=1024",
      "name": "Black Frill Tie Shoulder Bodycon Dress",
      "price": 7.99,
    },
    Object {
      "colour": "Red",
      "id": 5,
      "img": "https://cdn-img.prettylittlething.com/f/7/1/8/f718a4011ddf92f48aeefff6da0f475178694599_cly0842_1.jpg?imwidth=1024",
      "name": "Red Pin Stripe Belt T Shirt Dress",
      "price": 17,
    },
  ],
}
`;
